packages/loopring_v3/contracts/aux/access/LoopringIOExchangeOwner.sol

```diff
@@ -60,16 +52,28 @@ contract LoopringIOExchangeOwner is SelectorBasedAccessManager
     }
 
     function submitBlocksWithCallbacks(
+        uint    preconditionBlockNumber,
+        bytes32 preconditionBlockHash,
         ExchangeData.Block[] memory blocks,
         BlockCallback[]      memory callbacks
         )
         external
+        nonReentrant
     {
         require(
             hasAccessTo(msg.sender, SUBMITBLOCKS_SELECTOR) || open,
             "PERMISSION_DENIED"
         );
 
+        if (preconditionBlockNumber == 0) {
+            require(preconditionBlockHash == bytes32(0), "INVALID_PRECONDITION_BLOCK_HASH");
+        } else if (block.number.sub(preconditionBlockNumber) <= 256) {
+            require(
+                blockhash(preconditionBlockNumber) == preconditionBlockHash,
+                "INVALID_PRECONDITION"
+            );
+        }
+
```

Comment:
@Brechtpd is there a way optimize `IExchangeV3(target).submitBlocks(blocks);` using assembly to avoid unnecessary data copying?